### PR TITLE
Use Microsoft.NetFX20 package in ResultProviders

### PR DESCRIPTION
**Pull request by tmat**
Microsoft.NetFX20 package provides a reference assembly for mscorlib 2.0. Use it in ResultProviders instead of referencing mscorlib 2.0 installed on the machine. 

Also remove a reference to System.dll.

### CODE REVIEW

**Comment by tmat** [comment]
@amcasey @KevinH-MS

File: src/ExpressionEvaluator/CSharp/Source/ResultProvider/CSharpResultProvider.csproj [hunk 1]
```diff
@@ -12,22 +12,14 @@
...
-    <!--
-      As a convenience for developer machines without .NET 2.0 installed, we'll allow building against v4.*.
-      However, "real" builds must still target v2.0, and we always want to target 2.0 if it's present.
-    -->
-    <TargetFrameworkVersion Condition="Exists('$(WinDir)\Microsoft.NET\Framework\v2.0.50727\mscorlib.dll') OR '$(RealSignBuild)' == 'true'">v2.0</TargetFrameworkVersion>
-    <!--
-      We get warnings about missing XML comments for publicly visible members in Microsoft.CodeAnalysis.CSharp.SyntaxKind
-    -->
-    <NoWarn>1591</NoWarn>
+    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
```
Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ResultProvider/CSharpResultProvider.csproj]: Sanity check: it will still run on 2.0?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        Never = 1,
+        Advanced = 2
+    }
+
+    [AttributeUsage(AttributeTargets.All)]
+    internal sealed class EditorBrowsableAttribute : Attribute
+    {
+        public EditorBrowsableAttribute(EditorBrowsableState state) { }
+    }
+}
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs]: Newline

File: src/ExpressionEvaluator/VisualBasic/Source/ResultProvider/packages.config [hunk 3]
```diff
@@ -1,3 +1,4 @@
 <?xml version="1.0" encoding="utf-8"?>
 <packages>
-</packages>
+  <package id="Microsoft.NetFX20" version="1.0.0-rc2" targetFramework="net20" />  
+</packages>
```
Comment (amcasey) [inline src/ExpressionEvaluator/VisualBasic/Source/ResultProvider/packages.config]: Newline

**Comment by amcasey** [comment]
The parts I understood LGTM. :)

File: src/ExpressionEvaluator/CSharp/Source/ResultProvider/CSharpResultProvider.csproj (hunk 1 above)
Comment (tmat) [inline src/ExpressionEvaluator/CSharp/Source/ResultProvider/CSharpResultProvider.csproj]: Yes, this is only the version of tools that are used. The properties set below make sure that msbuild doesn't try to apply implicitly generated stuff for 4.5.

**Comment by tmat** [comment]
@jasonmalinowski to review the msbuild hackery :)

File: src/ExpressionEvaluator/CSharp/Source/ResultProvider/CSharpResultProvider.csproj (hunk 1 above)
Comment (jasonmalinowski) [inline src/ExpressionEvaluator/CSharp/Source/ResultProvider/CSharpResultProvider.csproj]: You might want to add a comment here that the TFV here is actually a lie.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
+﻿// A few dependencies from System.dll, which we want to avoid when compiling against Desktop Framework 2.0.
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs]: Copyright?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+namespace System.Linq
+{
+    internal class Dummy { }
```
Comment (jasonmalinowski) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs]: Why is this necessary? What's including System.Linq in the first place?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                throw new InvalidOperationException(message ?? "Assertion failed");
+            }
+        }
+    }
+}
+
+namespace System.CodeDom.Compiler
+{
+    [AttributeUsage(AttributeTargets.All, Inherited = false, AllowMultiple = false)]
+    internal sealed class GeneratedCodeAttribute : Attribute
```
Comment (jasonmalinowski) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs]: Why? Does the compiler get angry if this doesn't exist? Can you just not build against mscorlib only in 2.0 world with our current compiler?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/EvalResultDataItem.cs [hunk 7]
```diff
@@ -113,7 +113,6 @@ private static DkmEvaluationResultFlags GetFlags(DkmClrValue value)
 
         protected override void OnClose()
         {
-            Debug.WriteLine("Closing " + FullName);
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/EvalResultDataItem.cs]: Is this not Portable?

Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/EvalResultDataItem.cs]: Never mind, I see the issue.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
+﻿// A few dependencies from System.dll, which we want to avoid when compiling against Desktop Framework 2.0.
+namespace System.Linq
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs]: I would move this to the bottom and add a comment to the affect of "// Avoid use of Linq when compiling against newer versions of mscorlib (for portability)..."

File: src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs (hunk 4 above)
Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs]: I would change the comment to something like:

// A few dependencies from System.dll...we will avoid referencing the full library to promote portability...

File: src/ExpressionEvaluator/VisualBasic/Source/ResultProvider/BasicResultProvider.vbproj [hunk 9]
```diff
@@ -13,11 +13,12 @@
...
-      As a convenience for developer machines without .NET 2.0 installed, we'll allow building against v4.*.
-      However, "real" builds must still target v2.0, and we always want to target 2.0 if it's present.
-    -->
-    <TargetFrameworkVersion Condition="Exists('$(WinDir)\Microsoft.NET\Framework\v2.0.50727\mscorlib.dll') OR '$(RealSignBuild)' == 'true'">v2.0</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
+    <GenerateTargetFrameworkAttribute>false</GenerateTargetFrameworkAttribute>
+    <AddAdditionalExplicitAssemblyReferences>false</AddAdditionalExplicitAssemblyReferences>
+    <NoStdLib>true</NoStdLib>
+    <FrameworkPathOverride>..\..\..\..\..\packages\Microsoft.NetFX20.1.0.0-rc2\lib\net20</FrameworkPathOverride>
+    <NoStdLib>true</NoStdLib>
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/VisualBasic/Source/ResultProvider/BasicResultProvider.vbproj]: duplicate

File: src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs (hunk 6 above)
Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs]: It's needed for our designer generated resource files (but we want to avoid referencing the entirety of System.dll just to make the generate code behind compile).

**Comment by KevinH-MS** [comment]
LGTM :shipit:

**Comment by jasonmalinowski** [comment]
I'm not sure "looks good" is something I can really say given the abuse this is applying to the poor project system. So we'll go with :grimacing: instead.

File: src/ExpressionEvaluator/CSharp/Source/ResultProvider/CSharpResultProvider.csproj (hunk 1 above)
Comment (tmat) [inline src/ExpressionEvaluator/CSharp/Source/ResultProvider/CSharpResultProvider.csproj]: Done.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs (hunk 5 above)
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs]: This is just needed since we auto-import System.Linq in all VB projects. Maybe we shouldn't but that would be a bigger change.

File: src/ExpressionEvaluator/VisualBasic/Source/ResultProvider/packages.config (hunk 3 above)
Comment (tmat) [inline src/ExpressionEvaluator/VisualBasic/Source/ResultProvider/packages.config]: Fixed.

File: src/ExpressionEvaluator/VisualBasic/Source/ResultProvider/BasicResultProvider.vbproj (hunk 9 above)
Comment (tmat) [inline src/ExpressionEvaluator/VisualBasic/Source/ResultProvider/BasicResultProvider.vbproj]: Fixed.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs (hunk 2 above)
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs]: Fixed.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    [AttributeUsage(AttributeTargets.All)]
+    internal sealed class EditorBrowsableAttribute : Attribute
+    {
+        public EditorBrowsableAttribute(EditorBrowsableState state) { }
+    }
+}
+
+namespace System.Linq
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/SystemHelpers.cs]: More specific comment about VB auto importing this would be useful then.
